File: csharp/custom-set/CustomSet.cs
using System;
using System.Linq;

public class CustomSet
{
    private int[] _items;
    public CustomSet(params int[] values)
    {
        _items = values.Distinct().ToArray();
    }

    public CustomSet Add(int value)
    {
        return !Contains(value) ? new(_items.Append(value).ToArray()) : new(_items);
    }

    public bool Empty()
    {
        return _items.Length == 0;
    }

    public bool Contains(int value)
    {
        return _items.Contains(value);
    }

    public bool Subset(CustomSet right)
    {
        return _items.All(right.Contains);
    }

    public bool Disjoint(CustomSet right)
    {
        return !_items.Any(right.Contains);
    }

    public CustomSet Intersection(CustomSet right)
    {
        return new(_items.Where(x => right.Contains(x)).ToArray());
    }

    public CustomSet Difference(CustomSet right)
    {
        return new(_items.Where(x => !right.Contains(x)).ToArray());
    }

    public CustomSet Union(CustomSet right)
    {
        return new(_items.Union(right._items).ToArray());
    }

    public override bool Equals(object obj) => obj is CustomSet set && Equals(set);
    public bool Equals(CustomSet right) => _items.All(right.Contains) && right._items.All(Contains);
    public override int GetHashCode() => _items.GetHashCode();
}

File: fsharp/custom-set/CustomSet.fs
module CustomSet

type CustomSet (items: int seq) =
    member this.items = items |> Seq.distinct |> Seq.sort |> Seq.toList 

let empty = new CustomSet([])

let singleton value = new CustomSet([value])

let isEmpty (set : CustomSet) = List.isEmpty set.items

let size (set : CustomSet) = List.length set.items

let fromList list = new CustomSet(list)

let toList (set : CustomSet) = set.items

let contains value (set : CustomSet) = List.contains value set.items

let insert value (set : CustomSet) = new CustomSet(value :: set.items)

let union (left : CustomSet) (right : CustomSet) = new CustomSet(left.items @ right.items)

let intersection (left : CustomSet) (right : CustomSet) = new CustomSet(left.items |> List.filter (fun x -> List.contains x right.items))

let difference (left : CustomSet) (right : CustomSet) = new CustomSet(left.items |> List.filter (fun x -> not (List.contains x right.items)))

let isSubsetOf (left : CustomSet) (right : CustomSet) =  left.items |> Seq.forall (fun x -> contains x right)

let isDisjointFrom (left : CustomSet) (right : CustomSet) = left.items |> Seq.forall (fun x -> not (contains x right))

let isEqualTo (left : CustomSet) (right : CustomSet) = left.items = right.items